whelktool/scripts/2019/01/delete-spec-temporal-subjects-and-update-refs.groovy

```diff
@@ -0,0 +1,167 @@
+
+
+String termToUri(term) {
+    return SAO_URI + '/' + URLEncoder.encode(term.capitalize(), "UTF-8").replaceAll("\\+", "%20")
+}
+
+String getIdOfTerm(term) {
+    String uri
+    selectBySqlWhere("""
+            collection = 'auth' AND data#>>'{@graph,1,@type}' = 'Geographic' AND data#>>'{@graph,1,prefLabel}' = '${term}'
+    """) { doc ->
+        uri = doc.graph[1][ID]
+    }
+    return uri
+}
+
+boolean updateReference(work) {
+    def extractedTerms = []
+    def entitiesToMove = []
+    def termsToChange = SUBJECTS_TO_DELETE.values() as List
+
+    if (!work.subject) return
+
+    ListIterator iterSubj = work.subject.listIterator()
+    while (iterSubj.hasNext()) {
+        subj = iterSubj.next()
+
+        if (subj[TYPE] == COMPLEX_SUBJECT_TYPE && subj['inScheme'] && subj['inScheme'][ID] == SAO_URI
+            && subj.termComponentList.any{ it[TYPE] == TEMPSUB_TYPE} &&
+                subj.termComponentList.any{ termsToChange.contains(it[PREFLABEL])}) {
+
+            ListIterator iter = subj.termComponentList.listIterator()
+
+            while(iter.hasNext()) {
+                cpx_term = iter.next()
+                if (cpx_term[TYPE] == TEMPSUB_TYPE && termsToChange.contains(cpx_term[PREFLABEL].toLowerCase())) {
+                    if (!extractedTerms.contains(termToUri(cpx_term[PREFLABEL]))) {
+                        extractedTerms << termToUri(cpx_term[PREFLABEL])
+                    }
+                    iter.remove()
+                }
+            }
+
+            //Extract remaining entity in ComplexSubject, if only one remains
+            if (subj.termComponentList.size() == 1) {
+                if (subj.termComponentList.get(0)[ID]) {
+                    if (!extractedTerms.contains(subj.termComponentList.get(0)[ID])) {
+                        extractedTerms << subj.termComponentList.get(0)[ID]
+                    }
+                } else if (subj.termComponentList.get(0)[PREFLABEL]) {
+                    String newUri
+
+                    if (subj.termComponentList.get(0)[TYPE] == 'Geographic')
+                        newUri = getIdOfTerm(subj.termComponentList.get(0)[PREFLABEL].capitalize())
+                    else
+                        newUri = findCanonicalId(termToUri(subj.termComponentList.get(0)[PREFLABEL].capitalize()))
+
+                    if (newUri) {
+                        if (!extractedTerms.contains(newUri))
+                            extractedTerms << newUri
+                    } else {
+                        //Special treatment of geographical local subjects which shall be able to export to bib 651.
+                        //Otherwise follow convert to label to be able to export to bib 653
+                        String keyLabel = 'label'
+                        if (subj.termComponentList.get(0)[TYPE] == 'Geographic') { keyLabel = 'prefLabel' }
+
+                        if (!entitiesToMove.size() || entitiesToMove.any{ it[TYPE] != subj.termComponentList.get(0)[TYPE] ||
```

Comment:
Missing comparison? Or `entitiesToMove.empty` might be more clear?

Reply:
Also, I'd suggest moving out the `entitiesToMove.any` block from the `if` to make it easier to read.